sdk/nodejs/provisioning/docker-provision/image.ts

```diff
@@ -0,0 +1,237 @@
+				return `${ binPath }.exe`;
+			default:
+				return binPath;
+		}
+	}
+
+	/**
+	 * pullEngineSessionBin will retrieve Dagger binary from its docker image
+	 * and copy it to the local host.
+	 * This function automatically resolves host's platform to copy the correct
+	 * binary.
+	 */
+	private pullEngineSessionBin(engineSessionBinPath: string): void {
+		// Create a temporary bin file path
+		const tmpBinPath = path.join(
+			this.cacheDir,
+			`temp-${ this.ENGINE_SESSION_BINARY_PREFIX }`
+		);
+
+		const dockerRunArgs = [
+			'docker',
+			'run',
+			'--rm',
+			'--entrypoint',
+			'/bin/cat',
+			this.imageRef.Ref,
+			`/usr/bin/${ this.ENGINE_SESSION_BINARY_PREFIX }-${ os.platform() }-${ os.arch() }`,
+		];
+
+		try {
+			const fd = fs.openSync(tmpBinPath, 'w', 0o700);
+
+			execaCommandSync(dockerRunArgs.join(' '), {
+				stdout: fd,
+				stderr: 'pipe',
+				encoding: null,
+				// Kill the process if parent exit.
+				cleanup: true,
+				// Throw on error
+				reject: false,
+				timeout: 300000,
+			});
+
+			fs.closeSync(fd);
+			fs.renameSync(tmpBinPath, engineSessionBinPath);
+		} catch (e) {
+			fs.rmSync(tmpBinPath);
+
+			throw new Error(`failed to copy engine session binary: ${ e }`);
+		}
+
+		// Remove all temporary binary files
+		// Ignore current engine session binary or other files that have not be
+		// created by this SDK.
+		try {
+			const files = fs.readdirSync(this.cacheDir);
+			files.forEach((file) => {
+				const filePath = `${ this.cacheDir }/${ file }`;
+				if (filePath === engineSessionBinPath || !file.startsWith(this.ENGINE_SESSION_BINARY_PREFIX)) {
+					return;
+				}
+
+				fs.unlinkSync(filePath);
+			});
+		} catch (e) {
+			// Log the error but do not interrupt program.
+			console.error('could not clean up temporary binary files');
+		}
+	}
+
+	/**
+	 * runEngineSession execute the engine binary and set up a GraphQL client that
+	 * target this engine.
+	 */
+	private async runEngineSession(engineSessionBinPath: string, opts: ConnectOpts): Promise<Client> {
+		const engineSessionArgs = [ engineSessionBinPath, '--remote', `docker-image://${ this.imageRef.Ref }` ];
+
+		if (opts.Workdir) {
+			engineSessionArgs.push('--workdir', opts.Workdir);
+		}
+		if (opts.Project) {
+			engineSessionArgs.push('--project', opts.Project);
+		}
+
+		const { stdout, pid } = execaCommand(engineSessionArgs.join(' '),
+			{
+				stderr: opts.OutputLog || process.stderr,
+
+				// Kill the process if parent exit.
+				cleanup: true,
+			});
+
+		// Register PID to kill it later.
+		this.enginePid = pid;
+
+		const stdoutReader = readline.createInterface({
+			input: stdout!,
+		});
+
+		// Set a timeout of 10 seconds by default
+		// Do not call the function if port is successfully retrieved.
+		const timeoutFct = setTimeout(async () => this.Close(), opts.Timeout || 10000);
+
+		for await (const line of stdoutReader) {
+			// Read line as a port number
+			const port = parseInt(line);
+
+			clearTimeout(timeoutFct);
```

Comment:
I think there might be a race condition here where we can read the port successfully but then the timeout gets hit before we clear it. 

I don't really know the best way of handling this in nodejs, but I think you can use `promise.Race` for this: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/race#using_promise.race_to_implement_request_timeout

Reply:
btw this is not a blocker for this PR if it's not straightforward to fix, we just should address it before release

Reply:
Issue created to keep track on this (https://github.com/dagger/dagger/issues/3833)